pymtproto/paymentprotocol.go

```diff
@@ -0,0 +1,293 @@
+	// and `x509+sha1` which is also insecure. So `x509+sha256` it is.
+	if paymentRequest.PkiType == nil || *paymentRequest.PkiType != "x509+sha256" {
+		return nil, errors.New("payment request PkiType is not x509+sha256")
+	}
+
+	// Unmarshal the certificate object
+	certificateProto := new(payments.X509Certificates)
+	if err := proto.Unmarshal(paymentRequest.PkiData, certificateProto); err != nil {
+		return nil, err
+	}
+
+	// Parse the certificates
+	var certs []*x509.Certificate
+	for _, certBytes := range certificateProto.Certificate {
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, err
+		}
+		certs = append(certs, cert)
+	}
+
+	// If the certificate is expired or not valid yet we return and error
+	if !c.skipExpirationChecks {
+		if time.Now().After(certs[0].NotAfter) {
+			return nil, errors.New("certificate is expired")
+		}
+	}
+	if time.Now().Before(certs[0].NotBefore) {
+		return nil, errors.New("certificate is not valid yet")
+	}
+
+	// Now make sure the cert is signed by a valid certificate authority
+	roots := x509.NewCertPool()
+	roots.AddCert(certs[1])
+
+	opts := x509.VerifyOptions{
+		Roots: roots,
+	}
+	if _, err := certs[0].Verify(opts); err != nil {
+		return nil, err
+	}
+
+	// Verify the signature on the PaymentRequest object
+	signature := paymentRequest.Signature
+	paymentRequest.Signature = []byte{} // Zero out the signature for verification
+
+	serializedPaymentRequest, err := proto.Marshal(paymentRequest)
+	if err != nil {
+		return nil, err
+	}
+	if err := certs[0].CheckSignature(certs[0].SignatureAlgorithm, serializedPaymentRequest, signature); err != nil {
+		return nil, err
+	}
+
+	// Parse the payment details and build the response
+	paymentDetails := new(payments.PaymentDetails)
+	if err := proto.Unmarshal(paymentRequest.SerializedPaymentDetails, paymentDetails); err != nil {
+		return nil, err
+	}
+
+	pr := &PaymentRequest{
+		PayToName: certs[0].Subject.CommonName,
+	}
+
+	for _, out := range paymentDetails.Outputs {
+		// We're going to return an error here if they ask us to pay an unparsable
+		// address. This is kind of lame as we should be able to pay any script but
+		// our gRPC API currently only works with addresses for convenience so we wont
+		// be able to pay arbitrary scripts right now anyway.
+		_, addrs, _, err := txscript.ExtractPkScriptAddrs(out.Script, c.params)
+		if err != nil {
+			return nil, err
+		}
+		if out.Amount == nil {
+			return nil, errors.New("nil payment amount")
+		}
+		output := Output{
+			Address: addrs[0],
+			Amount:  bchutil.Amount(int64(*out.Amount)),
+		}
+		pr.Outputs = append(pr.Outputs, output)
+	}
+	if paymentDetails.Expires == nil {
+		return nil, errors.New("expiration time is nil")
+	}
+	pr.Expires = time.Unix(int64(*paymentDetails.Expires), 0)
+
+	if !c.skipExpirationChecks {
+		if pr.Expires.Before(time.Now()) {
+			return nil, errors.New("payment request is expired")
+		}
+	}
+
+	if paymentDetails.Memo != nil {
```

Comment:
Maybe use the getters here provided by Protobuf? paymentDetails.GetMemo() I believe.